docs/Protected_build_porting_guide.md

```diff
@@ -0,0 +1,363 @@
+EXTERN(umm_givesemaphore)
+
+EXTERN(malloc)
+EXTERN(realloc)
+EXTERN(zalloc)
+EXTERN(free)
+```
+- Then replace the assigned 'kernel' code memory with the 'user' code memory and 'kernel' data memory with the 'user' data memory.
+- When the userspace binary is created, we would like two additional things to appear as compared to the kernel space binary.
+- First, the userspace object needs to be placed at the starting of the userspace binary.
+- For example : For a board having flash memory, under the 'SECTIONS' command the changes are as follows :
+
+```
+	.userspace : {
+		*(.userspace)
+	} > uflash
+```
+
+- Second, a pointer to the user space heap needs to be placed at the start of the text section. This will be used by the mm (memory manager) module to make allocations.
+- Add user heap pointer (.usrheapptr) in the text section of the user memory.
+
+```
+.text : {
+	_stext = ABSOLUTE(.);
+	*(.usrheapptr)		/* User heap pointer added in the .text user section */
+	....
+	_etext = ABSOLUTE(.);
+	} > uflash
+```
+
+
+
+### Userspace Initialization
+- For the case of the separate user and kernel-space build, platform specific initialization of the user memory is required.
+- Normally this just means intializing the userspace .data and .bss segments
+- For this create a file "CHIPSET_userspace.c" in the folder path <os/arch/arm/src/CHIPSET> and define an API as show below to perform this action.
+- Also create a header file "CHIPSET_userspace.h" at path <os/arch/arm/src/CHIPSET> and declare the below given API.
+- The newly created file needs to be included into the Makefile for the CHIPSET (arch/arm/src/CHIPSET/Makefile)
+- This API clears all of user-space .bss and initializes all of user-space .data
+- This is done using the address information stored in the MACRO:'USERSPACE' (which is defined using CONFIG_TINYARA_USERSPACE)
+- The bss/data start and end addresses are stored in this macro.
+- Below is an example showing the basic implementation of this API :
+
+```
+void CHIPSET_userspace(void)
+	{
+		  ....
+		/* Clear all of user-space .bss */
+
+		dest = (uint8_t *)USERSPACE->us_bssstart;	/*STORING THE START ADDRESS OF .bss section */
+		end  = (uint8_t *)USERSPACE->us_bssend;		/*STORING THE END ADDRESS OF .bss section */
+
+		while (dest != end)
+		{
+		  *dest++ = 0;					/*Initialization*/
+		}
+
+		/* Initialize all of user-space .data */
+
+		src  = (uint8_t *)USERSPACE->us_datasource;	/*STORING THE VALUE WITH WHICH .data section is to be initialized with */
+		dest = (uint8_t *)USERSPACE->us_datastart;	/*STORING THE START ADDRESS OF .data section */
+		end  = (uint8_t *)USERSPACE->us_dataend;	/*STORING THE END ADDRESS OF .data section */
+
+		while (dest != end)
+		{
+		  *dest++ = *src++;				/*Initialization */
+		}
+		....
+	}
+```
+
+- This user initialization API(CHIPSET_userspace()) must called by the '__start()' API present in the "CHIPSET_start.c" (path : <os/arch/arm/src/chip> )
+
+
+
+### MPU Intialization
```

Comment:
Is it possible to provide standard hooks for MPU to make easy for porting across all hardware?

Reply:
Standard MPU api's are already provided in mpu.h file. 
But, based on each board and memory types and sizes, they mpu regions need to be configured using the api's in mpu.h..  This is explained in this section.